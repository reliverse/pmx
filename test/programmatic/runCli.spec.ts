import type { MockInstance } from "vitest";

import fs from "fs-extra";
import { tmpdir } from "node:os";
import path from "pathe";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";

import type { Runner } from "../../src";

import {
  AGENTS,
  parseNa,
  parseNi,
  parseNlx,
  parseNu,
  parseNun,
  runCli,
} from "../../src";

let basicLog: MockInstance;
let errorLog: MockInstance;
let warnLog: MockInstance;
let infoLog: MockInstance;

function runCliTest(
  fixtureName: string,
  agent: string,
  runner: Runner,
  args: string[],
) {
  return async () => {
    const cwd = await fs.mkdtemp(path.join(tmpdir(), "pmx-"));
    const fixture = path.join(__dirname, "..", "fixtures", fixtureName, agent);
    await fs.copy(fixture, cwd);

    await runCli(
      async (agent, _, ctx) => {
        // we override the args to be test specific
        return runner(agent, args, ctx);
      },
      {
        programmatic: true,
        cwd,
        args,
      },
    ).catch((e) => {
      // it will always throw if ezspawn is mocked
      if (e.command) {
        expect(e.command).toMatchSnapshot();
      } else {
        expect(e.message).toMatchSnapshot();
      }
    });
  };
}

beforeAll(() => {
  basicLog = vi.spyOn(console, "log");
  warnLog = vi.spyOn(console, "warn");
  errorLog = vi.spyOn(console, "error");
  infoLog = vi.spyOn(console, "info");

  vi.mock("tinyexec", async (importOriginal) => {
    const mod = (await importOriginal()) as any;
    return {
      ...mod,
      x: (cmd: string, args?: string[]) => {
        // break execution flow for easier snapshotting
        // eslint-disable-next-line no-throw-literal
        throw { command: [cmd, ...(args ?? [])].join(" ") };
      },
    };
  });
});

afterAll(() => {
  vi.resetAllMocks();
});

const agents = [...AGENTS, "unknown"];
const fixtures = ["lockfile", "packager"];

// matrix testing of: fixtures x agents x commands
fixtures.forEach((fixture) =>
  describe(fixture, () =>
    agents.forEach((agent) =>
      describe(agent, () => {
        /** pa */
        it("na", runCliTest(fixture, agent, parseNa, []));
        it("na run foo", runCliTest(fixture, agent, parseNa, ["run", "foo"]));

        /** pi */
        it("pi", runCliTest(fixture, agent, parseNi, []));
        it("pi foo", runCliTest(fixture, agent, parseNi, ["foo"]));
        it("pi foo -D", runCliTest(fixture, agent, parseNi, ["foo", "-D"]));
        it("pi --frozen", runCliTest(fixture, agent, parseNi, ["--frozen"]));
        it("pi -g foo", runCliTest(fixture, agent, parseNi, ["-g", "foo"]));

        /** plx */
        it("plx", runCliTest(fixture, agent, parseNlx, ["foo"]));

        /** pu */
        it("pu", runCliTest(fixture, agent, parseNu, []));
        it("pu -i", runCliTest(fixture, agent, parseNu, ["-i"]));

        /** pun */
        it("pun foo", runCliTest(fixture, agent, parseNun, ["foo"]));
        it("pun -g foo", runCliTest(fixture, agent, parseNun, ["-g", "foo"]));

        it("no logs", () => {
          expect(basicLog).not.toHaveBeenCalled();
          expect(warnLog).not.toHaveBeenCalled();
          expect(errorLog).not.toHaveBeenCalled();
          expect(infoLog).not.toHaveBeenCalled();
        });
      }),
    )),
);
