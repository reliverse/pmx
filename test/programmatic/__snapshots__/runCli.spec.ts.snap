// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lockfile > bun > pa 1`] = `"bun"`;

exports[`lockfile > bun > pa run foo 1`] = `"bun run foo"`;

exports[`lockfile > bun > pi --frozen 1`] = `"bun install --frozen-lockfile"`;

exports[`lockfile > bun > pi -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > bun > pi 1`] = `"bun install"`;

exports[`lockfile > bun > pi foo -D 1`] = `"bun add foo -d"`;

exports[`lockfile > bun > pi foo 1`] = `"bun add foo"`;

exports[`lockfile > bun > plx 1`] = `"bun x foo"`;

exports[`lockfile > bun > pu -i 1`] = `"bun update"`;

exports[`lockfile > bun > pu 1`] = `"bun update"`;

exports[`lockfile > bun > pun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > bun > pun foo 1`] = `"bun remove foo"`;

exports[`lockfile > npm > pa 1`] = `"npm"`;

exports[`lockfile > npm > pa run foo 1`] = `"npm run foo"`;

exports[`lockfile > npm > pi --frozen 1`] = `"npm ci"`;

exports[`lockfile > npm > pi -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > npm > pi 1`] = `"npm i"`;

exports[`lockfile > npm > pi foo -D 1`] = `"npm i foo -D"`;

exports[`lockfile > npm > pi foo 1`] = `"npm i foo"`;

exports[`lockfile > npm > nlx 1`] = `"npx foo"`;

exports[`lockfile > npm > nu -i 1`] = `"Command "upgrade-interactive" is not support by agent "npm""`;

exports[`lockfile > npm > nu 1`] = `"npm update"`;

exports[`lockfile > npm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > npm > nun foo 1`] = `"npm uninstall foo"`;

exports[`lockfile > pnpm > pa 1`] = `"pnpm"`;

exports[`lockfile > pnpm > pa run foo 1`] = `"pnpm run foo"`;

exports[`lockfile > pnpm > pi --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > pnpm > pi -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > pnpm > pi 1`] = `"pnpm i"`;

exports[`lockfile > pnpm > pi foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > pnpm > pi foo 1`] = `"pnpm add foo"`;

exports[`lockfile > pnpm > nlx 1`] = `"pnpm dlx foo"`;

exports[`lockfile > pnpm > nu -i 1`] = `"pnpm update -i"`;

exports[`lockfile > pnpm > nu 1`] = `"pnpm update"`;

exports[`lockfile > pnpm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > pnpm > nun foo 1`] = `"pnpm remove foo"`;

exports[`lockfile > pnpm@6 > pa 1`] = `"pnpm"`;

exports[`lockfile > pnpm@6 > pa run foo 1`] = `"pnpm run foo"`;

exports[`lockfile > pnpm@6 > pi --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > pnpm@6 > pi -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > pnpm@6 > pi 1`] = `"pnpm i"`;

exports[`lockfile > pnpm@6 > pi foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > pnpm@6 > pi foo 1`] = `"pnpm add foo"`;

exports[`lockfile > pnpm@6 > nlx 1`] = `"pnpm dlx foo"`;

exports[`lockfile > pnpm@6 > nu -i 1`] = `"pnpm update -i"`;

exports[`lockfile > pnpm@6 > nu 1`] = `"pnpm update"`;

exports[`lockfile > pnpm@6 > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > pnpm@6 > nun foo 1`] = `"pnpm remove foo"`;

exports[`lockfile > unknown > pa 1`] = `"pnpm"`;

exports[`lockfile > unknown > pa run foo 1`] = `"pnpm run foo"`;

exports[`lockfile > unknown > pi --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`lockfile > unknown > pi -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > unknown > pi 1`] = `"pnpm i"`;

exports[`lockfile > unknown > pi foo -D 1`] = `"pnpm add foo -D"`;

exports[`lockfile > unknown > pi foo 1`] = `"pnpm add foo"`;

exports[`lockfile > unknown > nlx 1`] = `"pnpm dlx foo"`;

exports[`lockfile > unknown > nu -i 1`] = `"pnpm update -i"`;

exports[`lockfile > unknown > nu 1`] = `"pnpm update"`;

exports[`lockfile > unknown > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > unknown > nun foo 1`] = `"pnpm remove foo"`;

exports[`lockfile > yarn > pa 1`] = `"yarn"`;

exports[`lockfile > yarn > pa run foo 1`] = `"yarn run foo"`;

exports[`lockfile > yarn > pi --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`lockfile > yarn > pi -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > yarn > pi 1`] = `"yarn install"`;

exports[`lockfile > yarn > pi foo -D 1`] = `"yarn add foo -D"`;

exports[`lockfile > yarn > pi foo 1`] = `"yarn add foo"`;

exports[`lockfile > yarn > nlx 1`] = `"npx foo"`;

exports[`lockfile > yarn > nu -i 1`] = `"yarn upgrade-interactive"`;

exports[`lockfile > yarn > nu 1`] = `"yarn upgrade"`;

exports[`lockfile > yarn > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > yarn > nun foo 1`] = `"yarn remove foo"`;

exports[`lockfile > yarn@berry > pa 1`] = `"yarn"`;

exports[`lockfile > yarn@berry > pa run foo 1`] = `"yarn run foo"`;

exports[`lockfile > yarn@berry > pi --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`lockfile > yarn@berry > pi -g foo 1`] = `"npm i -g foo"`;

exports[`lockfile > yarn@berry > pi 1`] = `"yarn install"`;

exports[`lockfile > yarn@berry > pi foo -D 1`] = `"yarn add foo -D"`;

exports[`lockfile > yarn@berry > pi foo 1`] = `"yarn add foo"`;

exports[`lockfile > yarn@berry > nlx 1`] = `"npx foo"`;

exports[`lockfile > yarn@berry > nu -i 1`] = `"yarn upgrade-interactive"`;

exports[`lockfile > yarn@berry > nu 1`] = `"yarn upgrade"`;

exports[`lockfile > yarn@berry > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`lockfile > yarn@berry > nun foo 1`] = `"yarn remove foo"`;

exports[`packager > bun > pa 1`] = `"bun"`;

exports[`packager > bun > pa run foo 1`] = `"bun run foo"`;

exports[`packager > bun > pi --frozen 1`] = `"bun install --frozen-lockfile"`;

exports[`packager > bun > pi -g foo 1`] = `"npm i -g foo"`;

exports[`packager > bun > pi 1`] = `"bun install"`;

exports[`packager > bun > pi foo -D 1`] = `"bun add foo -d"`;

exports[`packager > bun > pi foo 1`] = `"bun add foo"`;

exports[`packager > bun > nlx 1`] = `"bun x foo"`;

exports[`packager > bun > nu -i 1`] = `"bun update"`;

exports[`packager > bun > nu 1`] = `"bun update"`;

exports[`packager > bun > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > bun > nun foo 1`] = `"bun remove foo"`;

exports[`packager > npm > pa 1`] = `"npm"`;

exports[`packager > npm > pa run foo 1`] = `"npm run foo"`;

exports[`packager > npm > pi --frozen 1`] = `"npm ci"`;

exports[`packager > npm > pi -g foo 1`] = `"npm i -g foo"`;

exports[`packager > npm > pi 1`] = `"npm i"`;

exports[`packager > npm > pi foo -D 1`] = `"npm i foo -D"`;

exports[`packager > npm > pi foo 1`] = `"npm i foo"`;

exports[`packager > npm > nlx 1`] = `"npx foo"`;

exports[`packager > npm > nu -i 1`] = `"Command "upgrade-interactive" is not support by agent "npm""`;

exports[`packager > npm > nu 1`] = `"npm update"`;

exports[`packager > npm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > npm > nun foo 1`] = `"npm uninstall foo"`;

exports[`packager > pnpm > pa 1`] = `"pnpm"`;

exports[`packager > pnpm > pa run foo 1`] = `"pnpm run foo"`;

exports[`packager > pnpm > pi --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > pnpm > pi -g foo 1`] = `"npm i -g foo"`;

exports[`packager > pnpm > pi 1`] = `"pnpm i"`;

exports[`packager > pnpm > pi foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > pnpm > pi foo 1`] = `"pnpm add foo"`;

exports[`packager > pnpm > nlx 1`] = `"pnpm dlx foo"`;

exports[`packager > pnpm > nu -i 1`] = `"pnpm update -i"`;

exports[`packager > pnpm > nu 1`] = `"pnpm update"`;

exports[`packager > pnpm > nun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > pnpm > nun foo 1`] = `"pnpm remove foo"`;

exports[`packager > pnpm@6 > pa 1`] = `"pnpm"`;

exports[`packager > pnpm@6 > pa run foo 1`] = `"pnpm run foo"`;

exports[`packager > pnpm@6 > pi --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > pnpm@6 > pi -g foo 1`] = `"npm i -g foo"`;

exports[`packager > pnpm@6 > pi 1`] = `"pnpm i"`;

exports[`packager > pnpm@6 > pi foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > pnpm@6 > pi foo 1`] = `"pnpm add foo"`;

exports[`packager > pnpm@6 > plx 1`] = `"pnpm dlx foo"`;

exports[`packager > pnpm@6 > pu -i 1`] = `"pnpm update -i"`;

exports[`packager > pnpm@6 > pu 1`] = `"pnpm update"`;

exports[`packager > pnpm@6 > pun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > pnpm@6 > pun foo 1`] = `"pnpm remove foo"`;

exports[`packager > unknown > pa 1`] = `"pnpm"`;

exports[`packager > unknown > pa run foo 1`] = `"pnpm run foo"`;

exports[`packager > unknown > pi --frozen 1`] = `"pnpm i --frozen-lockfile"`;

exports[`packager > unknown > pi -g foo 1`] = `"npm i -g foo"`;

exports[`packager > unknown > pi 1`] = `"pnpm i"`;

exports[`packager > unknown > pi foo -D 1`] = `"pnpm add foo -D"`;

exports[`packager > unknown > pi foo 1`] = `"pnpm add foo"`;

exports[`packager > unknown > plx 1`] = `"pnpm dlx foo"`;

exports[`packager > unknown > pu -i 1`] = `"pnpm update -i"`;

exports[`packager > unknown > pu 1`] = `"pnpm update"`;

exports[`packager > unknown > pun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > unknown > pun foo 1`] = `"pnpm remove foo"`;

exports[`packager > yarn > pa 1`] = `"yarn"`;

exports[`packager > yarn > pa run foo 1`] = `"yarn run foo"`;

exports[`packager > yarn > pi --frozen 1`] = `"yarn install --frozen-lockfile"`;

exports[`packager > yarn > pi -g foo 1`] = `"npm i -g foo"`;

exports[`packager > yarn > pi 1`] = `"yarn install"`;

exports[`packager > yarn > pi foo -D 1`] = `"yarn add foo -D"`;

exports[`packager > yarn > pi foo 1`] = `"yarn add foo"`;

exports[`packager > yarn > plx 1`] = `"npx foo"`;

exports[`packager > yarn > pu -i 1`] = `"yarn upgrade-interactive"`;

exports[`packager > yarn > pu 1`] = `"yarn upgrade"`;

exports[`packager > yarn > pun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > yarn > pun foo 1`] = `"yarn remove foo"`;

exports[`packager > yarn@berry > pa 1`] = `"yarn"`;

exports[`packager > yarn@berry > pa run foo 1`] = `"yarn run foo"`;

exports[`packager > yarn@berry > pi --frozen 1`] = `"yarn install --immutable"`;

exports[`packager > yarn@berry > pi -g foo 1`] = `"npm i -g foo"`;

exports[`packager > yarn@berry > pi 1`] = `"yarn install"`;

exports[`packager > yarn@berry > pi foo -D 1`] = `"yarn add foo -D"`;

exports[`packager > yarn@berry > pi foo 1`] = `"yarn add foo"`;

exports[`packager > yarn@berry > plx 1`] = `"yarn dlx foo"`;

exports[`packager > yarn@berry > pu -i 1`] = `"yarn up -i"`;

exports[`packager > yarn@berry > pu 1`] = `"yarn up"`;

exports[`packager > yarn@berry > pun -g foo 1`] = `"npm uninstall -g foo"`;

exports[`packager > yarn@berry > pun foo 1`] = `"yarn remove foo"`;
